services/net/reporting/ReportingManager.cs

```diff
@@ -0,0 +1,442 @@
+    /// </summary>
+    /// <param name="sender"></param>
+    /// <param name="e"></param>
+    /// <returns>True if the consumer should retry the message.</returns>
+    private void ListenerErrorHandler(object sender, ErrorEventArgs e)
+    {
+        // Only the first retry will count as a failure.
+        if (_retries == 0)
+            this.State.RecordFailure();
+
+        if (e.GetException() is ConsumeException consume)
+        {
+            if (consume.Error.IsFatal)
+                this.Listener.Stop();
+        }
+    }
+
+    /// <summary>
+    /// The Kafka consumer has stopped which means we need to also cancel the background task associated with it.
+    /// </summary>
+    /// <param name="sender"></param>
+    /// <param name="e"></param>
+    private void ListenerStopHandler(object sender, EventArgs e)
+    {
+        if (_consumer != null &&
+            !_notRunning.Contains(_consumer.Status) &&
+            _cancelToken != null && !_cancelToken.IsCancellationRequested)
+        {
+            _cancelToken.Cancel();
+        }
+    }
+
+    /// <summary>
+    /// Retrieve a file from storage and send to Microsoft Cognitive Services. Obtain
+    /// the report and update the content record accordingly.
+    /// </summary>
+    /// <param name="result"></param>
+    /// <returns></returns>
+    private async Task HandleMessageAsync(ConsumeResult<string, ReportRequestModel> result)
+    {
+        try
+        {
+            // The service has stopped, so to should consuming messages.
+            if (this.State.Status != ServiceStatus.Running)
+            {
+                this.Listener.Stop();
+                this.State.Stop();
+            }
+            else
+            {
+                await ProcessReportAsync(result);
+
+                // Inform Kafka this message is completed.
+                this.Listener.Commit(result);
+                this.Listener.Resume();
+
+                // Successful run clears any errors.
+                this.State.ResetFailures();
+                _retries = 0;
+            }
+        }
+        catch (Exception ex)
+        {
+            if (ex is HttpClientRequestException httpEx)
+            {
+                this.Logger.LogError(ex, "HTTP exception while consuming. {response}", httpEx.Data["body"] ?? "");
+            }
+            else
+            {
+                this.Logger.LogError(ex, "Failed to handle message");
+            }
+            ListenerErrorHandler(this, new ErrorEventArgs(ex));
+        }
+    }
+
+    /// <summary>
+    /// Process the report request.
+    /// </summary>
+    /// <param name="request"></param>
+    /// <returns></returns>
+    private async Task ProcessReportAsync(ConsumeResult<string, ReportRequestModel> result)
+    {
+        var request = result.Message.Value;
+        if (request.Destination.HasFlag(ReportDestination.ReportingService))
+        {
+            if (request.ReportInstanceId.HasValue)
```

Comment:
If an `ReportInstanceId` is provided it means we're either sending a custom report or resending an existing report.